File: Code/Arrays_Bit_Manipulation/Container_With_Most_Water.go
package main

import (
    "fmt"
)

func maxArea(height []int) int {
    left, right := 0, len(height)-1
    maxWater := 0

    for left < right {
        h := min(height[left], height[right])
        w := right - left
        area := h * w
        if area > maxWater {
            maxWater = area
        }

        if height[left] < height[right] {
            left++
        } else {
            right--
        }
    }

    return maxWater
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func main() {
    height := []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
    fmt.Println("Maximum water that can be contained:", maxArea(height))
}

File: Code/Arrays_Bit_Manipulation/Container_With_Most_Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let left = 0, right = height.length - 1;
    let maxWater = 0;

    while (left < right) {
        let h = Math.min(height[left], height[right]);
        let w = right - left;
        let area = h * w;
        maxWater = Math.max(maxWater, area);

        if (height[left] < height[right])
            left++;
        else
            right--;
    }

    return maxWater;
};

// Test case
const height = [1, 8, 6, 2, 5, 4, 8, 3, 7];
const result = maxArea(height);
console.log("Maximum water that can be contained:", result);
